src/test/lib/HandshakeTest.cpp

```diff
@@ -885,24 +891,452 @@ QuicTestVersionNegotiation(
                 TEST_TRUE(Client.IsValid());
 
                 TEST_QUIC_SUCCEEDED(
-                    Client.SetQuicVersion(168430090ul)); // Random reserved version to force VN.
+                    Client.Start(
+                        ClientConfiguration,
+                        QuicAddrFamily,
+                        QUIC_LOCALHOST_FOR_AF(QuicAddrFamily),
+                        ServerLocalAddr.GetPort()));
+                if (!Client.WaitForConnectionComplete()) {
+                    return;
+                }
+
+                TEST_TRUE(Client.GetIsConnected());
+                TEST_TRUE(Client.GetStatistics().VersionNegotiation);
+                TEST_EQUAL(Client.GetQuicVersion(), LATEST_SUPPORTED_VERSION);
+            }
+        }
+    }
+}
+
+void
+QuicTestCompatibleVersionNegotiation(
+    _In_ int Family
+    )
+{
+    uint32_t ClientVersions[] = { QUIC_VERSION_1_MS_H, QUIC_VERSION_1_H };
+    uint32_t ServerVersions[] = { QUIC_VERSION_1_H, QUIC_VERSION_1_MS_H };
+    const uint32_t ClientVersionsLength = ARRAYSIZE(ClientVersions);
+    const uint32_t ServerVersionsLength = ARRAYSIZE(ServerVersions);
+    const uint32_t ExpectedResultVersion = QUIC_VERSION_1_H;
+
+    MsQuicSettings ClientSettings;
+    ClientSettings.SetDesiredVersionsList(ClientVersions, ClientVersionsLength);
+    ClientSettings.SetIdleTimeoutMs(3000);
+
+    MsQuicSettings ServerSettings;
+    ServerSettings.SetDesiredVersionsList(ServerVersions, ServerVersionsLength);
+    ServerSettings.SetIdleTimeoutMs(3000);
+
+    MsQuicSettings ClearVNSettings;
+    ClearVNSettings.SetDesiredVersionsList(nullptr, 0);
+
+    TEST_QUIC_SUCCEEDED(
```

Comment:
The server settings being global makes an unavoidable test problem: the server settings will persist across test cases if one of the version negotiation tests that set the server version fails.

Reply:
We'll likely need to create an RAII wrapper for any tests that modify the global VN data, so on error or passing it will get reset back to stock.

Reply:
gtest has a nifty thing where I could put all the VN tests into their own test class witch a cleanup function run on the end of every test case, regardless of pass/fail. The problem exists primarily for kernel mode tests, where the driver doesn't unload.

Reply:
Yeah we can do that too. Either that or just any time we set it in a test we do it through a wrapper struct. Both would end up working about the same.

Reply:
But how does the RAII wrapper work for kernel? The test doesn't throw exceptions in kernel; all the macros just set a global to false and return immediately.

Reply:
When it returns, the stack will still unwind just like a normal function return. So it will call all destructors for stack allocated variables. Nothing special would need to happen, since that happens on returns anyway.